<a href="https://colab.research.google.com/github/feernandarodg/feernandarodg/blob/main/EXPO_2_0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1.Libraries**

In [ ]:
import numpy as np
import pandas as pd
import plotly.express as px

# **2. Import File**

* Upload the file to Colab
* Right click on the file and copy the path


In [ ]:
# read the file

## Save the path
p = '/content/Economic.xlsx'

# Read ans create the dataframe
d = pd.read_excel(p)
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 7 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Country Name                   1300 non-null   object 
 1   Country Code                   1300 non-null   object 
 2   Year                           1300 non-null   int64  
 3   GDP Per Capita                 1300 non-null   float64
 4   Education Expenditure (% GDP)  1300 non-null   float64
 5   Health Expenditure (% GDP)     1300 non-null   float64
 6   R&D                            1300 non-null   float64
dtypes: float64(4), int64(1), object(2)
memory usage: 71.2+ KB


In [ ]:
# Explore
d.head()

Country Name Country Code  Year  GDP Per Capita  \
0      Albania          ALB  2013     4413.063397   
1      Albania          ALB  2014     4578.633208   
2      Albania          ALB  2015     3952.803584   
3      Albania          ALB  2016     4124.055390   
4      Albania          ALB  2017     4531.032207   

   Education Expenditure (% GDP)  Health Expenditure (% GDP)           R&D  
0                       12.12391                    6.272344  1.543943e+07  
1                       12.74000                    6.441643  1.598555e+07  
2                       11.17755                    6.483612  1.376044e+07  
3                       13.59696                    6.666062  1.433367e+07  
4                       12.39080                    6.546046  1.573369e+07

In [ ]:
# Explore
d.tail()

Country Name Country Code  Year  GDP Per Capita  \
1295       Zambia          ZMB  2018     1475.199836   
1296       Zambia          ZMB  2019     1268.120941   
1297       Zambia          ZMB  2020      956.831747   
1298       Zambia          ZMB  2021     1137.344395   
1299       Zambia          ZMB  2022     1487.907764   

      Education Expenditure (% GDP)  Health Expenditure (% GDP)           R&D  
1295                      17.118719                    5.056766  1.785288e+07  
1296                      15.291870                    5.319501  1.581540e+07  
1297                      12.378020                    5.617884  1.228843e+07  
1298                      11.514140                    5.200252  1.502762e+07  
1299                      10.447814                    5.200252  2.020935e+07

# **3.Cleaning**

## Column names

In [ ]:
# Country Name to Country
d.rename(columns = {'Country Name':'Country'}, inplace = True)

# Country Code to Code
d.rename(columns = {'Country Code':'Code'}, inplace = True)

# GDP Per Capita to GDPPC
d.rename(columns = {'GDP Per Capita':'GDPPC'}, inplace = True)

# Education Expenditure (% GDP) to EducationEx
d.rename(columns = {'Education Expenditure (% GDP)':'EducationEx'}, inplace = True)

# Health Expenditure (% GDP) to HealthEx
d.rename(columns = {'Health Expenditure (% GDP)':'HealthEx'}, inplace = True)

d.head()


Country Code  Year        GDPPC  EducationEx  HealthEx           R&D
0  Albania  ALB  2013  4413.063397     12.12391  6.272344  1.543943e+07
1  Albania  ALB  2014  4578.633208     12.74000  6.441643  1.598555e+07
2  Albania  ALB  2015  3952.803584     11.17755  6.483612  1.376044e+07
3  Albania  ALB  2016  4124.055390     13.59696  6.666062  1.433367e+07
4  Albania  ALB  2017  4531.032207     12.39080  6.546046  1.573369e+07

# **4. Indexing**

In [ ]:
# Duplicate the dataframe
d2 = d

In [ ]:
# Duplicate year as string / how to create a column
d2['Year_str'] = d2['Year'].astype(str)
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Country      1300 non-null   object 
 1   Code         1300 non-null   object 
 2   Year         1300 non-null   int64  
 3   GDPPC        1300 non-null   float64
 4   EducationEx  1300 non-null   float64
 5   HealthEx     1300 non-null   float64
 6   R&D          1300 non-null   float64
 7   Year_str     1300 non-null   object 
dtypes: float64(4), int64(1), object(3)
memory usage: 81.4+ KB


In [ ]:
# Indexing
idx = pd.MultiIndex.from_product([d2['Country'].unique(), range(d2.Year.min(), d2.Year.max()+1)],
                                 names = ['Country', 'Year'])

## Attach the index
d2 = d2.set_index(['Country', 'Year'])

d2.head()

Code        GDPPC  EducationEx  HealthEx           R&D Year_str
Country Year                                                                
Albania 2013  ALB  4413.063397     12.12391  6.272344  1.543943e+07     2013
        2014  ALB  4578.633208     12.74000  6.441643  1.598555e+07     2014
        2015  ALB  3952.803584     11.17755  6.483612  1.376044e+07     2015
        2016  ALB  4124.055390     13.59696  6.666062  1.433367e+07     2016
        2017  ALB  4531.032207     12.39080  6.546046  1.573369e+07     2017

In [ ]:
d2['GDPPC'].unique()

array([4413.06339739, 4578.63320812, 3952.80358411, ...,  956.83174749,
       1137.34439483, 1487.90776431])

##**5 Delete NaN**

In [ ]:
# Display if there ara NaN
d2.isna()

Code  GDPPC  EducationEx  HealthEx    R&D  Year_str
Country Year                                                      
Albania 2013  False  False        False     False  False     False
        2014  False  False        False     False  False     False
        2015  False  False        False     False  False     False
        2016  False  False        False     False  False     False
        2017  False  False        False     False  False     False
...             ...    ...          ...       ...    ...       ...
Zambia  2018  False  False        False     False  False     False
        2019  False  False        False     False  False     False
        2020  False  False        False     False  False     False
        2021  False  False        False     False  False     False
        2022  False  False        False     False  False     False

[1300 rows x 6 columns]

In [ ]:
# Delete NaN
d2 = d2.dropna()
d2.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 1300 entries, ('Albania', 2013) to ('Zambia', 2022)
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Code         1300 non-null   object 
 1   GDPPC        1300 non-null   float64
 2   EducationEx  1300 non-null   float64
 3   HealthEx     1300 non-null   float64
 4   R&D          1300 non-null   float64
 5   Year_str     1300 non-null   object 
dtypes: float64(4), object(2)
memory usage: 70.3+ KB


In [ ]:
# By the flies
d2 = d2.replace('..',np.NaN)
d2.head()

Code        GDPPC  EducationEx  HealthEx           R&D Year_str
Country Year                                                                
Albania 2013  ALB  4413.063397     12.12391  6.272344  1.543943e+07     2013
        2014  ALB  4578.633208     12.74000  6.441643  1.598555e+07     2014
        2015  ALB  3952.803584     11.17755  6.483612  1.376044e+07     2015
        2016  ALB  4124.055390     13.59696  6.666062  1.433367e+07     2016
        2017  ALB  4531.032207     12.39080  6.546046  1.573369e+07     2017

# **6. Descriptive Stadistics**

In [ ]:
# Basic
d2 [['GDPPC','EducationEx','HealthEx', 'R&D']].describe().applymap(lambda x: '{:.1f}'.format(x))

GDPPC EducationEx HealthEx             R&D
count    1300.0      1300.0   1300.0          1300.0
mean    17238.3        14.3      6.5   12442882926.9
std     22159.6         4.5      2.6   59607629517.8
min       216.8         4.6      1.8        624596.8
25%      2597.7        10.8      4.4      41128688.3
50%      6960.5        13.6      6.3     282308747.5
75%     22362.9        17.2      8.3    2977121712.8
max    133590.1        38.1     18.8  726624259812.8

ANALYSIS: There is a great deal of variation in GDPPC, EducationEx, HealthEx, and R&D expenditure across countries.
The median values for each of these variables are higher than the mean values, indicating that there are a few countries with very high values that are skewing the mean.
The high standard deviations for each of these variables also indicate that there is a great deal of variation within each country. This means that there are some people within each country who are much wealthier, healthier, and more educated than others.
The high median R&D expenditure is particularly noteworthy. This suggests that many countries are investing heavily in research and development, which could lead to new technologies and innovations that improve the lives of people around the world.
For example, the high variation in GDPPC suggests that there is a great deal of inequality between countries. The high median R&D expenditure suggests that many countries are investing in the future, which is a positive sign. However, the high variation in EducationEx and HealthEx suggests that there are still many people in the world who do not have access to quality education and healthcare.

In [ ]:
# By indexd2
d2[['EducationEx']].groupby('Country').describe().applymap(lambda x:'{:.1f}'.format(x))

EducationEx                                         
                    count  mean  std   min   25%   50%   75%   max
Country                                                           
Albania              10.0  12.2  0.9  11.0  11.7  12.1  12.7  13.6
Algeria              10.0  15.9  1.2  13.8  15.2  16.2  16.5  17.6
Angola               10.0   6.9  1.1   5.4   6.2   6.6   6.9   8.9
Argentina            10.0  13.5  1.0  11.9  12.7  13.6  14.3  14.6
Armenia              10.0   9.9  1.0   8.3   9.1  10.1  10.6  11.4
...                   ...   ...  ...   ...   ...   ...   ...   ...
United States        10.0  14.1  1.2  12.6  13.2  13.7  14.4  16.0
Uruguay              10.0  15.1  0.6  14.4  14.7  15.0  15.3  16.4
Uzbekistan           10.0  22.5  1.4  20.5  21.8  22.4  22.9  25.6
Vietnam              10.0  15.8  1.4  14.0  14.9  15.5  16.6  18.5
Zambia               10.0  14.9  2.8  10.4  13.0  15.3  16.2  20.1

[130 rows x 8 columns]

ANALYSIS: The data presented in the image show that there is a large variation in the education level between countries. Countries with the highest levels of education are in Western Europe and North America, with an average of more than 14 years of schooling. Countries with the lowest levels of education are in Sub-Saharan Africa and South Asia, with an average of less than 9 years of schooling.

These data are indicative of the inequalities that exist in access to education globally. Countries with high levels of education have a competitive advantage in the labor market and are more likely to achieve economic development.

To address this problem, countries must take steps to ensure that all children have access to quality education. This includes investing in school infrastructure, improving teacher training, and providing support to students who need additional help.

## **7. Basic Graph**

In [ ]:
# Scatter matrix
fig1 = px.scatter_matrix(d2, dimensions = ['GDPPC','EducationEx', 'HealthEx', 'R&D',], color = 'Year_str')
fig1.update_traces(diagonal_visible = False)
fig1.update_traces(showupperhalf = False)
fig1.show()

ANALYSIS:
The plot shows that there is a positive correlation between GDP per capita and all three of the other variables. This means that countries with higher GDP per capita tend to have higher education expenditure, health expenditure, and R&D expenditure.

The plot also shows that the relationship between GDP per capita and the other variables has strengthened over time. This is evident from the fact that the red dots (representing the later years) are more concentrated in the upper right corner of the plot than the blue dots (representing the earlier years).

Overall, the plot suggests that there is a strong relationship between GDP per capita and investment in education, health, and R&D. This relationship has strengthened over time, suggesting that countries are increasingly recognizing the importance of investing in these areas.

The correlation between GDP per capita and education expenditure is the strongest, followed by the correlation between GDP per capita and R&D expenditure. The correlation between GDP per capita and health expenditure is the weakest. The relationship between GDP per capita and the other variables is not linear. There is a greater increase in education expenditure, health expenditure, and R&D expenditure as GDP per capita increases at higher levels. There is a large variation in education expenditure, health expenditure, and R&D expenditure among countries with similar GDP per capita. This suggests that there are other factors, such as cultural values and government policies, that also influence investment in these areas.

## **8. Pivots**

Filter

In [ ]:
china = d2.filter(like = 'China', axis = 0)
china.head()

Code        GDPPC  EducationEx  HealthEx           R&D Year_str
Country Year                                                                
China   2013  CHN  7020.385682     13.48708  4.714907  1.912046e+11     2013
        2014  CHN  7636.074223     12.85351  4.777906  2.118622e+11     2014
        2015  CHN  8016.446016     12.10193  4.938280  2.275377e+11     2015
        2016  CHN  8094.390375     11.88548  4.979556  2.359367e+11     2016
        2017  CHN  8817.045496     11.58869  5.065922  2.604937e+11     2017

In [ ]:
y20 = d2.filter(like = '2020', axis=0)
y20.head()

Code        GDPPC  EducationEx   HealthEx           R&D  \
Country   Year                                                           
Albania   2020  ALB  5343.037704    12.070000   6.093392  1.832341e+07   
Algeria   2020  DZA  3354.157303    16.549191   6.321180  3.781661e+08   
Angola    2020  AGO  1502.950754     6.467230   2.911835  1.622294e+07   
Argentina 2020  ARG  8496.428157    11.916900   9.984381  1.950185e+09   
Armenia   2020  ARM  4505.867746     8.827780  12.241035  2.644770e+07   

               Year_str  
Country   Year           
Albania   2020     2020  
Algeria   2020     2020  
Angola    2020     2020  
Argentina 2020     2020  
Armenia   2020     2020

*Pivots*

In [ ]:
## BE CAREFUL!!! YOU NEED TO USE AN UNINDEXED DATA FRAME
# Example 1
### rows    = Years
### columns = Code
### Values  = GDPPC

pivot_GDPPC = d.pivot(index = 'Year',
                      columns = 'Code',
                      values = 'GDPPC')
pivot_GDPPC.head()

Code          AGO          ALB           ARE           ARG          ARM  \
Year                                                                      
2013  5101.983876  4413.063397  45729.607676  13080.254732  3833.157071   
2014  5059.080441  4578.633208  46865.964598  12334.798245  4017.229913   
2015  3100.830685  3952.803584  41525.138904  13789.060425  3666.141825   
2016  1709.515534  4124.055390  41054.539570  12790.264064  3679.952349   
2017  2283.214233  4531.032207  43063.967479  14613.035715  4041.995072   

Code           AUS           AUT          AZE         BDI           BEL  ...  \
Year                                                                     ...   
2013  68156.386105  50731.127254  7875.344611  241.547666  46757.951856  ...   
2014  62515.314832  51786.377175  7890.835250  257.818557  47764.071512  ...   
2015  56708.961197  44195.817595  5500.504620  289.359627  41008.296719  ...   
2016  49876.712376  45307.587862  3880.686267  242.539527  42012.622719  ...   
2017  53934.154374  47429.158456  4147.198619  244.145422  44198.482391  ...   

Code         TZA         UGA          UKR           URY           USA  \
Year                                                                    
2013  927.454901  819.757867  4187.739845  18140.891918  53291.127689   
2014  983.277882  897.509729  2948.917876  18131.578847  55123.849787   
2015  901.714089  864.180059  2016.009571  16950.753376  56762.729452   
2016  914.933020  753.684406  2074.359455  16837.940380  57866.744934   
2017  946.841626  766.177604  2500.282534  18995.397020  59907.754261   

Code          UZB          VCT          VNM          ZAF          ZMB  
Year                                                                   
2013  2419.718744  7117.554763  2367.499542  7441.230854  1840.320553  
2014  2628.460054  7210.616549  2558.778924  6965.137897  1724.576220  
2015  2753.971058  7386.737484  2595.234979  6204.929901  1307.909649  
2016  2704.677189  7684.779282  2760.717101  5735.066787  1249.923143  
2017  1916.764625  7996.656233  2992.071746  6734.475153  1495.752138  

[5 rows x 130 columns]

# **9. Another Graphs**

## (One) Line graph

In [ ]:
france = d2.filter(like = 'France', axis = 0)
france.head()

Code         GDPPC  EducationEx   HealthEx           R&D Year_str
Country Year                                                                  
France  2013  FRA  42602.717965     9.611372  11.404669  6.290253e+10     2013
        2014  FRA  43068.548724     9.634183  11.534149  6.499946e+10     2014
        2015  FRA  36652.922305     9.620544  11.440772  5.432122e+10     2015
        2016  FRA  37062.533572     9.559823  11.485308  5.495866e+10     2016
        2017  FRA  38781.049487     9.648250  11.357558  5.706426e+10     2017

In [ ]:
fig1 = px.line(france, x = 'Year_str', y = 'HealthEx')
fig1.show()

In [ ]:
# Without the Legend
fig1.update_layout(showlegend = False)


# Center the title
fig1.update_layout(title_text = 'Public spending on health France',
                  title_font_size = 30,
                  title_x = 0.5)

# Modify the text in the axis
fig1.update_xaxes(title = 'Year', title_font_size = 25,
                 dtick = 'M1',
                 tickformat = '%b\n%Y')
fig1.update_yaxes(title = 'USD', title_font_size = 25)

fig1.show()

ANALYSIS: The graph shows public health spending in France per year, expressed in USD. Data is available from 2013 to 2022.

Public health spending in France has increased steadily in recent years, rising from USD 108.2 billion in 2013 to USD 122.2 billion in 2022. This increase is due to a number of factors, including They include the aging of the population, the increase in the prevalence of chronic diseases and the development of new medical technologies.

Public health spending represents a significant proportion of total French government spending. In 2022, public health spending represented 11.4% of France's GDP. This is more than the European Union average, which is 9.8% of GDP.

The graph also shows that public health spending in France has been affected by the COVID-19 pandemic. In 2020, public health spending increased by 8.5%, due to increased demand for health care during the pandemic.

To summarize, public spending on health in France is higher than the European Union average. This is due to a number of factors, including the French social security system, which provides universal coverage to all citizens, and the high cost of healthcare in France.
Public spending on health in France has increased faster than GDP. This is because the cost of healthcare is rising faster than inflation.

## Double axis

In [ ]:
## libraries
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [ ]:
## Create a figure with secondary axis

fig3 = make_subplots(specs = [[{'secondary_y': True}]])

## Add each graph
fig3.add_trace(
    go.Scatter(x = france.Year_str, y = france.GDPPC, name = 'GDPPC.'),
    secondary_y = False,
)

## Add each graph
fig3.add_trace(
    go.Scatter(x = france.Year_str, y = france.EducationEx, name = 'EducationEx.'),
    secondary_y = True,
)


ANALYSIS: In general, it is observed that there is a positive relationship between both variables. As public spending on education increases, GDP per capita also increases. This relationship is stronger in the first years of the period, when public spending on education increases significantly.
This positive relationship between public spending on education and GDP per capita can be explained by several factors. First of all, education is a key factor for economic development. A quality education system can help improve the productivity of the workforce, which would lead to an increase in GDP.
Second, public spending on education can help reduce poverty and inequality. People with higher education are more likely to find a well-paying job and escape poverty.
It is important to know that the relationship between public spending on education and GDP per capita may not be as direct as seen in the graph. In particular, it is possible that the relationship is stronger for certain levels of public spending on education.
For example, public spending on education is likely to have a significant impact on GDP per capita when it exceeds a certain threshold. This is because as public spending on education increases, students have access to higher quality education, which can lead to an increase in workforce productivity and GDP.
However, public spending on education may not have a significant impact on GDP per capita when it is below a certain threshold. This is because, with a low level of public spending on education, students may have access to a basic education, but may not have the skills and knowledge necessary to contribute fully to the economy.

In [ ]:
## Create a figure with secondary axis

fig4 = make_subplots(specs = [[{'secondary_y': True}]])

## Add each graph
fig4.add_trace(
    go.Bar(x = france.Year_str, y = france.GDPPC, name = 'GDPPC.'),
    secondary_y = False,
)

## Add each graph
fig4.add_trace(
    go.Scatter(x = france.Year_str, y = france.EducationEx, name = 'EducationEx.'),
    secondary_y = True,
)


## Pie Chart

In [ ]:
### Sort the data frame
y20_rd = y20.sort_values('GDPPC', ascending=False)

### Extract 30 countries
y20_rd_25 = y20_rd.iloc[0:25]

In [ ]:
fig5 = px.pie(y20_rd_25, values = 'GDPPC', names = y20_rd_25.index.get_level_values(0))
fig5.show()

STORY:

## Box Plot

In [ ]:
fig6 = px.box(d2, x = 'Year_str', y = 'EducationEx')
fig6.show()

linea dentro de la caja es la media
- menor a
+ mayor a

Historia

##10. A little more complex graph

In [ ]:
## Create a list to select countries
c_list = ['Colombia', 'Chile', 'Peru', 'Mexico', 'Brazil', 'Argentina']

## Select from the dataframe (unidexed)
d3 = d2.reset_index()
latam = d3[d3['Country'].isin(c_list)]

latam.head()

Country  Year Code         GDPPC  EducationEx   HealthEx           R&D  \
30  Argentina  2013  ARG  13080.254732     14.45793   9.780979  3.414220e+09   
31  Argentina  2014  ARG  12334.798245     13.79933   9.671300  3.126128e+09   
32  Argentina  2015  ARG  13789.060425     13.96331  10.229337  3.703028e+09   
33  Argentina  2016  ARG  12790.264064     13.35554   9.001898  3.111867e+09   
34  Argentina  2017  ARG  14613.035715     13.26392  10.442257  3.580569e+09   

   Year_str  
30     2013  
31     2014  
32     2015  
33     2016  
34     2017

In [ ]:
## GRAPH
fig7 = px.line(latam, x = 'HealthEx', y = 'GDPPC', text = 'Year_str', color = 'Country')
fig7.update_traces(textposition = 'top center')
fig7.show()



STORYhbjjbkjnknk